pkg/graphql/execution_engine.go

```diff
@@ -0,0 +1,113 @@
+package graphql
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/jensneuse/abstractlogger"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/execution"
+	"github.com/jensneuse/graphql-go-tools/pkg/execution/datasource"
+	"github.com/jensneuse/graphql-go-tools/pkg/operationreport"
+)
+
+type DataSourceHttpJsonOptions struct {
+	HttpClient *http.Client
+}
+
+type DataSourceGraphqlOptions struct {
+	HttpClient *http.Client
+}
+
+type ExecutionOptions struct {
+	ExtraArguments json.RawMessage
+}
+
+type ExecutionEngine struct {
+	logger      abstractlogger.Logger
+	basePlanner *datasource.BasePlanner
+	executor    *execution.Executor
+	schema      *Schema
+}
+
+func NewExecutionEngine(logger abstractlogger.Logger, schema *Schema, plannerConfig datasource.PlannerConfiguration) (*ExecutionEngine, error) {
+	executor := execution.NewExecutor(nil)
+	basePlanner, err := datasource.NewBaseDataSourcePlanner(schema.rawInput, plannerConfig, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ExecutionEngine{
+		logger:      logger,
+		basePlanner: basePlanner,
+		executor:    executor,
+		schema:      schema,
+	}, nil
+}
+
+func (e *ExecutionEngine) AddHttpJsonDataSource(name string) error {
+	return e.AddHttpJsonDataSourceWithOptions(name, DataSourceHttpJsonOptions{})
+}
+
+func (e *ExecutionEngine) AddHttpJsonDataSourceWithOptions(name string, options DataSourceHttpJsonOptions) error {
+	httpJsonFactoryFactory := &datasource.HttpJsonDataSourcePlannerFactoryFactory{}
+
+	if options.HttpClient != nil {
+		httpJsonFactoryFactory.Client = options.HttpClient
+	}
+
+	return e.AddDataSource(name, httpJsonFactoryFactory)
+}
+
+func (e *ExecutionEngine) AddGraphqlDataSource(name string) error {
+	return e.AddGraphqlDataSourceWithOptions(name, DataSourceGraphqlOptions{})
+}
+
+func (e *ExecutionEngine) AddGraphqlDataSourceWithOptions(name string, options DataSourceGraphqlOptions) error {
+	graphqlFactoryFactory := &datasource.GraphQLDataSourcePlannerFactoryFactory{}
+
+	if options.HttpClient != nil {
+		graphqlFactoryFactory.Client = options.HttpClient
+	}
+
+	return e.AddDataSource(name, graphqlFactoryFactory)
+}
+
+func (e *ExecutionEngine) AddDataSource(name string, plannerFactoryFactory datasource.PlannerFactoryFactory) error {
+	return e.basePlanner.RegisterDataSourcePlannerFactory(name, plannerFactoryFactory)
+}
+
+func (e *ExecutionEngine) Execute(ctx context.Context, operation *Request, writer io.Writer) error {
```

Comment:
Instead of giving an io.Writer, I implemented a proposal here: https://github.com/TykTechnologies/graphql-go-tools/pull/30

The idea is not giving just a buffer to user instead providing a http response getter. We need http response on gateway and also if there will be some users using this api, they will mostly need to convert the buffer into http response. 

Usage will be as simple as this:
```
var result *graphql.ExecutionResult
result, err = p.TykAPISpec.graphQLEngine.ExecuteWithOptions(context.Background(), &gqlRequest, graphql.ExecutionOptions{ExtraArguments: nil})
res=result.GetAsHTTPResponse()
```

Reply:
If user doesn't want http response, give it like this and let him work on it:
`result.Buffer()`

Reply:
Buffer/Write Control should be on the client side - this is not possible when the lib creates its own buffer.

Reply:
Let's not sign conversation as resolved to let other reviewers see it.